lib/driver.js

```diff
@@ -244,23 +281,47 @@ class EspressoDriver extends BaseDriver {
       }
     };
 
-    const isApk = _.endsWith(_.toLower(appPath), '.apk');
-    // Do not cache .apk if was already present on the local file system
-    const shouldResultAppPathBeCached = !isApk || (isApk && isUrl);
+    const hasApkExt = (appPath) => _.endsWith(_.toLower(appPath), APK_EXT);
+    const hasZipkExt = (appPath) => _.endsWith(_.toLower(appPath), ZIP_EXT);
+    const extractUniversalApk = async (shouldExtract, appPath) =>
+      shouldExtract ? appPath : await this.adb.extractUniversalApk(appPath);
+
     let pathInCache = null;
     let isResultAppPathAlreadyCached = false;
     if (_.isPlainObject(cachedAppInfo)) {
       const packageHash = await fs.hash(appPath);
       if (packageHash === cachedAppInfo.packageHash && await fs.exists(cachedAppInfo.fullPath)) {
+        logger.info(`Using '${cachedAppInfo.fullPath}' that is cached from '${appPath}'`);
         isResultAppPathAlreadyCached = true;
         pathInCache = cachedAppInfo.fullPath;
       }
     }
+
+    // appPath can be .zip, .apk, .apks or .aab
+    const isApk = hasApkExt(appPath);
+    // local .zip, .aab and downloaded .zip, .aab and .apks should be cached
+    // to skip unzip and generate .apk from .aab.
+    const shouldResultAppPathBeCached = !isApk || (isApk && isUrl);
+
     if (!isResultAppPathAlreadyCached) {
       if (shouldResultAppPathBeCached) {
-        pathInCache = isApk ? appPath : await this.adb.extractUniversalApk(appPath);
+        // .zip, .aab or downloaded .apk
+        const unzippedAppPath = await this.unzipAppIfNeeded(appPath);
+        const isUnzippedAppPath = !!unzippedAppPath;
+
+        // Probably the appPath zip file did not have SUPPORTED_EXTENSIONS.
+        if (hasZipkExt(appPath) && !isUnzippedAppPath) {
+          logger.errorAndThrow(`${this.opts.app} did not have any of '${SUPPORTED_EXTENSIONS.join(', ')}' ` +
+            `extension packages. Please make sure the zip file has a valid package.`);
+        }
+
+        // unzippedAppPath or appPath has SUPPORTED_EXTENSIONS.
+        pathInCache = isUnzippedAppPath
+          ? await extractUniversalApk(hasApkExt(unzippedAppPath), unzippedAppPath)
+          : await extractUniversalApk(isApk, appPath);
+
         if (!isApk && isUrl) {
```

Comment:
I assume the clenup won't work as expected if .zip archive contains .aab file. According to the current logic the .aab  will be extracted, then transformed to .apk, but it won't be cleaned up. Only the parent .zip is going to be deleted if it was not a local file in such case

Reply:
Yes, you're correct. b9b348a cleans the local .aap that is unzipped one